ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/StateRef.java

```diff
@@ -70,11 +70,18 @@ public int getIndex() {
      */
     public static StateRef parse(@NotNull final String value, DigestService digestService) {
         try {
+            if (value.replaceAll("[^" + DELIMITER + "]", "").length() != 2) {
```

Comment:
I'm not sure this is safe - `replaceAll` performs [in-line replacement of the string](https://docs.oracle.com/javase/8/docs/api/java/lang/String.html#replaceAll-java.lang.String-java.lang.String-), so this will have a side effect on `value`, which might cause issues if it is used again in the calling code.

[This](https://www.baeldung.com/java-count-chars#4-using-java-8-features) may be a safer and cleaner approach. Unfortunately there doesn't seem to be an easy way with the standard Java libraries to do this.

Alternatively, I question whether it makes sense to have all this implementation in `corda-api`, or if we should create a `StateRefImpl` in `corda-runtime-os`, much as we do with `SecureHash` and `SecureHashImpl`. That means we'd have access to slightly nicer Kotlin libs. But that would be a much bigger change and we'd need to be careful not to break backwards compatibility.

@vlajos @nkovacsx any thoughts on the above? My feeling is maybe we'd have the separate impl if we were starting again, but it's a bit of a risk to do it now?

Reply:
> I'm not sure this is safe - `replaceAll` performs [in-line replacement of the string](https://docs.oracle.com/javase/8/docs/api/java/lang/String.html#replaceAll-java.lang.String-java.lang.String-), so this will have a side effect on `value`, which might cause issues if it is used again in the calling code.
> 
> [This](https://www.baeldung.com/java-count-chars#4-using-java-8-features) may be a safer and cleaner approach. Unfortunately there doesn't seem to be an easy way with the standard Java libraries to do this.
> 
> Alternatively, I question whether it makes sense to have all this implementation in `corda-api`, or if we should create a `StateRefImpl` in `corda-runtime-os`, much as we do with `SecureHash` and `SecureHashImpl`. That means we'd have access to slightly nicer Kotlin libs. But that would be a much bigger change and we'd need to be careful not to break backwards compatibility.
> 
> @vlajos @nkovacsx any thoughts on the above? My feeling is maybe we'd have the separate impl if we were starting again, but it's a bit of a risk to do it now?

I think it is a bit late now.
Also, I think it is good to have this parse accessible without more services.